pkg/components/cluster-autoscaler/component_internal_test.go

```diff
@@ -0,0 +1,319 @@
+					Code: "doh",
+				},
+			},
+		},
+	}
+
+	if w := getClusterWorkers("", device.Facility.Code, d); len(w) != 2 {
+		t.Fatalf("got workers from other facility: %+v", w)
+	}
+}
+
+func TestGetClusterWorkersFilterByCluster(t *testing.T) {
+	clusterName := "baz"
+	facility := &packngo.Facility{
+		Code: "bar",
+	}
+
+	device := packngo.Device{
+		DeviceRaw: packngo.DeviceRaw{
+			Hostname: fmt.Sprintf("%s-worker", clusterName),
+			Facility: facility,
+		},
+	}
+
+	d := []packngo.Device{
+		device,
+		device,
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "doh-worker",
+				Facility: facility,
+			},
+		},
+	}
+
+	if w := getClusterWorkers(clusterName, device.Facility.Code, d); len(w) != 2 {
+		t.Fatalf("got workers from other cluster: %+v", w)
+	}
+}
+
+func TestGetClusterWorkersFilterNonWorkers(t *testing.T) {
+	facility := &packngo.Facility{
+		Code: "bar",
+	}
+
+	device := packngo.Device{
+		DeviceRaw: packngo.DeviceRaw{
+			Hostname: "worker",
+			Facility: facility,
+		},
+	}
+
+	d := []packngo.Device{
+		device,
+		device,
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "controller",
+				Facility: facility,
+			},
+		},
+	}
+
+	if w := getClusterWorkers("", device.Facility.Code, d); len(w) != 2 {
+		t.Fatalf("got devices which are not workers: %+v", w)
+	}
+}
+
+func TestFindDuplicatedDevicesNonUniqueHostname(t *testing.T) {
+	device := packngo.Device{
+		DeviceRaw: packngo.DeviceRaw{
+			Hostname: "foo",
+		},
+	}
+
+	d := []packngo.Device{
+		device,
+		device,
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "bar",
+			},
+		},
+	}
+
+	if w := findDuplicatedDevices(d); len(w) != 1 {
+		t.Fatalf("two devices with the same hostname should be treated as duplicates")
+	}
+}
+
+func TestFindDuplicatedDevicesUniqueHostname(t *testing.T) {
+	d := []packngo.Device{
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "bar",
+			},
+		},
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "foo",
+			},
+		},
+	}
+
+	if w := findDuplicatedDevices(d); len(w) != 0 {
+		t.Fatalf("two devices with the different hostname should not be treated as duplicates")
+	}
+}
+
+func TestDeviceHostnames(t *testing.T) {
+	d := []packngo.Device{
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "bar",
+			},
+		},
+		{
+			DeviceRaw: packngo.DeviceRaw{
+				Hostname: "foo",
+			},
+		},
+	}
+
+	expected := []string{"bar", "foo"}
+
+	if hostnames := devicesHostnames(d); !reflect.DeepEqual(expected, hostnames) {
+		t.Fatalf("expected: %+v, got: %+v", expected, hostnames)
+	}
+}
+
+func TestGetWorkerUserdataNoUserdataOnError(t *testing.T) {
+	if userdata, _ := getWorkerUserdata("", "", []packngo.Device{}); userdata != "" {
+		t.Fatalf("if error is returned, no userdata should be returned either")
+	}
+}
+
+func TestGetWorkerUserdataNoWorkers(t *testing.T) {
+	if _, err := getWorkerUserdata("", "", []packngo.Device{}); err == nil {
+		t.Fatalf("if there is no devices to get user data from, error should be returned")
+	}
+}
```

Comment:
Aren't these two same tests? Since the test input is same I think both can be coupled into one test.

Reply:
Yes, it should be fine to merge them.

Reply:
Merged.